torchrl/data/utils.py

```diff
@@ -35,6 +42,98 @@
 INDEX_TYPING = Union[None, int, slice, str, Tensor, List[Any], Tuple[Any, ...]]
 
 
+def consolidate_spec(
+    spec: TensorSpec,
+    recurse_through_entries: bool = True,
+    recurse_through_stack: bool = True,
+):
+    """Given a TensorSpec, removes exclusive keys by adding 0 shaped specs."""
+    spec = spec.clone()
+
+    if not isinstance(spec, (CompositeSpec, LazyStackedCompositeSpec)):
+        return spec
+
+    if isinstance(spec, LazyStackedCompositeSpec):
+        keys = set(spec.keys())  # shared keys
+        lazy_keys_per_spec = [
+            set() for _ in range(len(spec._specs))
+        ]  # list of exclusive keys per td
+        lazy_keys_examples = {}  # set of all exclusive keys with an example for each
+        for spec_index in range(len(spec._specs)):  # gather all lazy keys
+            sub_spec = spec._specs[spec_index]
+            if recurse_through_stack:
+                sub_spec = consolidate_spec(
+                    sub_spec, recurse_through_entries, recurse_through_stack
+                )
+                spec._specs[spec_index] = sub_spec
+            for sub_spec_key in sub_spec.keys():
+                if sub_spec_key not in keys:  # lazy key
+                    lazy_keys_per_spec[spec_index].add(sub_spec_key)
+                    if sub_spec_key not in lazy_keys_examples:
+                        value = sub_spec[sub_spec_key]
+                        lazy_keys_examples.update({sub_spec_key: value})
+
+        for spec_index in range(len(spec._specs)):  # add missing exclusive entries
+            sub_spec = spec._specs[spec_index]
```

Comment:
- it is not items because here i am cycling through the list, not the values. there are still missing keys which will not be present in keys(). the values are checked later when we have made all keys non-exclusive.
- we are indexing with integers because i want to `spec._specs[spec_index] = sub_spec`